ouroboros-network-framework/src/Ouroboros/Network/Snocket.hs

```diff
@@ -192,6 +192,8 @@ newtype TestAddress addr = TestAddress { getTestAddress :: addr }
   deriving (Eq, Ord, Generic, Typeable)
   deriving Show via Quiet (TestAddress addr)
 
+instance Hashable addr => Hashable (TestAddress addr)
```

Comment:
Because we have constraints such as:

```haskell
, HasHeader header
, HasHeader block
, HeaderHash header ~ HeaderHash block
```

which will instantiate to `NtNAddr` types

Reply:
Sorry, I forgot this one. NtNAddr is not part of the header, but this is the type error I get if I remove the `Hashable` instance for `TestAddr addr`:

```haskell
test/Test/Ouroboros/Network/Diffusion/Node.hs:260:7: error:
    • Could not deduce (hashable-1.3.5.0:Data.Hashable.Class.Hashable
                          NtNAddr)
        arising from a use of ‘blockFetchLogic’
      from the context: (MonadAsync m, MonadEvaluate m, MonadFix m,
                         MonadFork m, MonadLabelledSTM m, MonadTraceSTM m, MonadMask m,
                         MonadST m, MonadTime m, MonadTimer m, MonadThrow (STM m),
                         resolver ~ (), forall a. Semigroup a => Semigroup (m a),
                         Eq (Async m Void))
        bound by the type signature for:
                   run :: forall resolver (m :: * -> *).
                          (MonadAsync m, MonadEvaluate m, MonadFix m, MonadFork m,
                           MonadLabelledSTM m, MonadTraceSTM m, MonadMask m, MonadST m,
                           MonadTime m, MonadTimer m, MonadThrow m, MonadThrow (STM m),
                           resolver ~ (), forall a. Semigroup a => Semigroup (m a),
                           Eq (Async m Void)) =>
                          Node.BlockGeneratorArgs Block StdGen
                          -> Node.LimitsAndTimeouts Block
                          -> Interfaces m
                          -> Arguments m
                          -> Diff.P2P.TracersExtra
                               NtNAddr
                               NtNVersion
                               NtNVersionData
                               NtCAddr
                               NtCVersion
                               NtCVersionData
                               ResolverException
                               m
                          -> m Void
        at test/Test/Ouroboros/Network/Diffusion/Node.hs:(148,1)-(173,13)
    • In a stmt of a 'do' block:
        blockFetchLogic
          nullTracer nullTracer (blockFetchPolicy blockHeapVar nodeKernel)
          registry
          (BlockFetchConfiguration
             {bfcMaxConcurrencyBulkSync = 1, bfcMaxConcurrencyDeadline = 2,
              bfcMaxRequestsInflight = 10, bfcDecisionLoopInterval = 0.01,
              bfcSalt = 0})
      In the expression:
        do blockHeapVar <- LazySTM.newTVarIO Set.empty
           blockFetchLogic
             nullTracer nullTracer (blockFetchPolicy blockHeapVar nodeKernel)
             registry
             (BlockFetchConfiguration
                {bfcMaxConcurrencyBulkSync = 1, bfcMaxConcurrencyDeadline = 2,
                 bfcMaxRequestsInflight = 10, bfcDecisionLoopInterval = 0.01,
                 bfcSalt = 0})
      In an equation for ‘blockFetch’:
          blockFetch registry nodeKernel
            = do blockHeapVar <- LazySTM.newTVarIO Set.empty
                 blockFetchLogic
                   nullTracer nullTracer (blockFetchPolicy blockHeapVar nodeKernel)
                   registry
                   (BlockFetchConfiguration
                      {bfcMaxConcurrencyBulkSync = 1, bfcMaxConcurrencyDeadline = 2,
                       bfcMaxRequestsInflight = 10, bfcDecisionLoopInterval = 0.01,
                       bfcSalt = 0})
    |
260 |       blockFetchLogic
    |       ^^^^^^^^^^^^^^^...
```

I do not really understand why `blockFetchLogic` requires `Hashable NtNAddr`, nothing in the constraints really requires it

Reply:
That's because of [`comparePeerGSV`](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/src/Ouroboros/Network/BlockFetch/DeltaQ.hs?plain=1#L40-L69=) function.

We have to be careful to not use `ordered-containers`.  We could use randomness, but I think we wanted to have something stable, hence the use of hash with a salt.